enhancements/split-sdk-into-multiple-projects.md

```diff
@@ -0,0 +1,221 @@
+---
+title: split-sdk-into-multiple-projects
+authors:
+  - "@jesusr"
+reviewers:
+  - "@estroz"
+  - "@hasbro17"
+  - "@joelanford"
+  - "@fabianvf"
+  - "@shawn-hurley"
+approvers:
+  - "@estroz"
+  - "@hasbro17"
+  - "@joelanford"
+  - "@fabianvf"
+  - "@shawn-hurley"
+creation-date: 2020-05-18
+last-updated: 2020-05-27
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Split SDK into multiple projects
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [X] Test plan is defined
+- [X] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+1. What do we do with other shared tooling, CI, test helpers, etc. (e.g.
+   changelog generator)? Do we need a separate repo for that?
+   * Answer: We should visit these on a case by case basis as we see what is needed
+     during the splitting.
+
+## Summary
+
+This is the overarching enhancement that will plan how to split functionality
+out of the current operator-sdk project into separate projects. The initial
+thought on component breakdown mirrors the kubebuilder and controller-runtime
+projects:
+
+1. Common libraries and operator building blocks
+1. Project scaffolding and the main SDK CLI
+1. Helm operator library and scaffolding plugin
+1. Ansible operator library and scaffolding plugin
+
+## Motivation
+
+The motivation is to avoid a breaking change in one component causing a major
+version bump for all other unrelated components. Another motivation is to make
+the projects consumable by other projects as libraries. A benefit would be
+better testing and easier to contribute to the smaller projects.
+
+While there are a few complexities with having separate repos, like PR
+coordination between repos, or just the working complexity of having to point
+at different repos to pick up your dep changes, these are obstacles we can
+overcome. They also aren’t that much different from other dependencies we
+work with.
+
+### Goals
+
+Split the operator-sdk repo into 4 separate repos:
+
+1. CLI and Go Operator
+1. SDK library
+1. Helm operator
+1. Ansible operator
+
+This SDK library split would be a requirement for a 1.0 release. The Ansible
+and Helm operators could go afterwards if need be.
+
+### Non-Goals
+
+* We will not solve the downstream SDK in this enhancement.
+* We will not solve all of the items required for a 1.0 release.
+* Defining when version bumps occur in each repo.
+
+## Proposal
+
+Split the operator-sdk repo into 4 separate repos as outlined in the Goals
+section above.
+
+1. CLI and Go Operator
+   * contains the Go operator code
+   * contains the CLI code
+   * related scaffolding
+   * lives in operator-framework/operator-sdk
+   * will have a deprecation policy
+     * we can deprecate in minor versions; removal becomes a major version bump
+   * this split is dependent on Kubebuilder Phase 2.
```

Comment:
I don't think this is true, since Helm/Ansible plugins are phase 1 plugins.

Reply:
I made it phase 2 because I didn't want the moving of repos to affect phase 2 work since that might touch a couple repos vs one right now.

Reply:
I updated the comment to be a bit more clear:

> the split of the CLI/Go Operator repo is dependent on Kubebuilder Phase 2
to avoid creating a moving target.

And I added this to the helm and ansible splits respectively:
> the Helm plugin is dependent on Kubebuilder Phase 1.
> the Ansible plugin is dependent on Kubebuilder Phase 1.